### DISCUSSION TITLE
Python Virtual Environment - Prompt Prepended with virtual env name

### DISCUSSION THREAD

**Discussion by jiaulislam** [JanDeDobbeleer/oh-my-posh#390] [Troubleshoot] [answered]
Hi. Thanks guys your works are awesome. 

Could anyone please help with below marked text to disable I just want the Theme Virtual env Prompt nothing else.

Terminal: Windows Terminal
Powershell version: 7.1.1
Font: Meslo S

![image](https://user-images.githubusercontent.com/54698236/107349559-8fac5d80-6af2-11eb-97c9-57f68cc53a54.png)

**Comment by lnu** [comment]
Did you follow the install guide to set the prompt in your powershell profile? What does it look like?

**Reply by jiaulislam** [reply]
@lnu thanks for your reply. 
I did followed the instructions from installation guide. But I was unable to fix that. As I'm not that pro with programming knowledge. However just the Time Date is showing but the Environ Variable is still there.

![image](https://user-images.githubusercontent.com/54698236/107485098-89c78280-6bad-11eb-8958-ae97c44066c9.png)

**Comment by JanDeDobbeleer** [comment, marked as answer]
@jiaulislam that's controlled by what's in the activate script. If you look at its content, there should be a switch to disable prepending the prompt (they hijack the current prompt). Depending on the type of virtual env you either need `export PYENV_VIRTUALENV_DISABLE_PROMPT=1`, `VIRTUAL_ENV_DISABLE_PROMPT=1` or something different. Best to look at what's in the script on your end ;-)

**Reply by jiaulislam** [reply]
![image](https://user-images.githubusercontent.com/54698236/109104302-f1590280-7755-11eb-8fd8-917f31844021.png)
It works now. Thanks for your time ❤️

**Reply by doraeric** [reply]
If `VIRTUAL_ENV_DISABLE_PROMPT=1` is set, it seems like there is no way to know whether the shell is in an activated environment.

**Reply by doraeric** [reply]
It turns out the default theme `jandedobbeleer` disabled `fetch_virtual_env`, so the prompt is not showing python venv.
Other themes like `agnoster` show the venv correctly, works fine with `VIRTUAL_ENV_DISABLE_PROMPT=1`

**Reply by AndrewLaganaro** [reply]
In Windows it calls for the activate script:

`C:\Users\Andrew\ProjectABC\.venv\Scripts\Activate.ps1`

inside of it, you need to find:


```
# Now set the environment variable VIRTUAL_ENV, used by many tools to determine
# that there is an activated venv.
$env:VIRTUAL_ENV = $VenvDir

if (-not $Env:VIRTUAL_ENV_DISABLE_PROMPT) {

    Write-Verbose "Setting prompt to '$Prompt'"

    # Set the prompt to include the env name
    # Make sure _OLD_VIRTUAL_PROMPT is global
    function global:_OLD_VIRTUAL_PROMPT { "" }
    Copy-Item -Path function:prompt -Destination function:_OLD_VIRTUAL_PROMPT
    New-Variable -Name _PYTHON_VENV_PROMPT_PREFIX -Description "Python virtual environment prompt prefix" -Scope Global -Option ReadOnly -Visibility Public -Value $Prompt
    # ## HERE IS WHERE THE ENV NAME IS SET ON THE TERMINAL. THE FOREGROUND COLOR IS THE MAIN INDICATOR THAT THIS IS WHAT IM LOOKING FOR
    # ## BY SETTING IT AS "" IT WILL NOT SHOW UP ON THE TERMINAL, BUT IT WILL STILL BE SET
    function global:prompt {
        Write-Host -NoNewline -ForegroundColor Green ""#"($_PYTHON_VENV_PROMPT_PREFIX) "
        _OLD_VIRTUAL_PROMPT
    }
    $env:VIRTUAL_ENV_PROMPT = $Prompt
}
```

As you can see I've added the comments under # ## (on my VScode with Better Comments extension it turns orange, to highlight them from the crowd. It can be customized to your liking)

And here:` ""#"($_PYTHON_VENV_PROMPT_PREFIX) "`

I've commented the env name that would show up in the terminal `#"($_PYTHON_VENV_PROMPT_PREFIX) "` and substituted it by a empty string ""

Then it wont bug anymore for me. Before when the env name showed up my rightside customization would be expelled to the line below. Removing it completely did the trick, and also my customization already shows the name of the Env I'm in because I configured it like this, as I'm a python user, so it was redundant anyway!

as you can see here, in the final result:
![image](https://github.com/user-attachments/assets/6314a68c-f1c2-417b-ac98-17bc82b39f02)

**Reply by AndrewLaganaro** [reply]
@JanDeDobbeleer huge thanks for choosing my comment from six months ago, I was actually looking for it again but github has no resource to search for comments hahah

Also, good lord. I'll make it better tomorrow with better formating, pretty useful to this day. Thanks again

**Comment by crow-rojas** [comment]
EDIT: I didn't read well the reply for @JanDeDobbeleer. I was trying to put the env. variable VIRTUAL_ENV_DISABLE_PROMPT in the PW profile but I had to place it in the activate ps1 file of the venv folder. But I managed to make this universal for any new virtualenv created and not only the virtualenv of an actual project. Just have to put VIRTUAL_ENV_DISABLE_PROMPT in ```C:\Users\<user>\AppData\Local\Programs\Python\Python39\Lib\venv\scripts\common\Activate.ps1```, after ```$env:VIRTUAL_ENV = $VenvDir```, like this:

```
...
# Now set the environment variable VIRTUAL_ENV, used by many tools to determine
# that there is an activated venv.
$env:VIRTUAL_ENV = $VenvDir
$env:VIRTUAL_ENV_DISABLE_PROMPT = 1

if (-not $Env:VIRTUAL_ENV_DISABLE_PROMPT) {
...
```
I Hope it helps for anyone!

---

Hi! Where do I have to put VIRTUAL_ENV_DISABLE_PROMPT=1? I putted in my ps1 profile file after the theme import but it didn't work. Like this:

```$env:VIRTUAL_ENV_DISABLE_PROMPT=1```

@JanDeDobbeleer @jiaulislam 

Thanks in advance :D

**Reply by Antyos** [reply]
I was able to get `$env:VIRTUAL_ENV_DISABLE_PROMPT = 1` when placing it in the default PowerShell `profile.ps1`. However, here are two things of note:
1. You have to restart the shell to refresh the profile (or manually refresh it) if you want the updated environment variables.
2. If you set the environment variable manually while in an active virtual environment, you have to deactivate/reactivate it.

**Reply by julioMayorga** [reply]
@JanDeDobbeleer I actually did all of the steps to get it working and I am unable to see the venv name in the theme. Am I missing a step? I am using python3.15

**Comment by ghost** [comment]
Hello, awesome work, 
I am using larserikfinholt.omp.json theme, I would like to add virtualenv
but don't know where to put this code from the doc.
{
  "type": "python",
  "style": "powerline",
  "powerline_symbol": "",
  "foreground": "#100e23",
  "background": "#906cff",
  "template": "  {{ .Full }} "
}

Thank you

**Reply by JanDeDobbeleer** [reply]
@wereket78 depending on how you're currently using oh-my-posh, you might need to first export your config. The customize section in the [docs](https://ohmyposh.dev/docs/installation/customize) has all that information.

**Reply by ghost** [reply]
I added a section after exporting the config, I modified the colors to match doom-moonlight, but the python env does not show after adding the section
https://github.com/wereket78/dotfiles/blob/main/executable_dot_mytheme.omp.json#L34

**Reply by JanDeDobbeleer** [reply]
You're missing the property to tell it to fetch the virtual env and the `.Venv` property in the `template`. Read the python [documentation](https://ohmyposh.dev/docs/segments/python).

**Comment by ghost** [comment]
Hello
I added property 
https://github.com/wereket78/dotfiles/blob/main/executable_dot_mytheme.omp.json#L39
It does not show the venv
thank you

**Reply by ghost** [reply]
It works now

**Comment by robinmordasiewicz** [comment]
conda config --set changeps1 False
http://conda.pydata.org/docs/config.html#change-command-prompt-changeps1

**Comment by PhyAMR** [comment]
![image](https://github.com/user-attachments/assets/a9aa8b04-b51e-4835-a5d2-5c2679540c85)
Hi, I'm using kushal theme and it shows the last venv activated, I really don't know when I'm on a virtual or not. I'm creating virtual enviroments with vscode and using a bash terminal

Is there any solution to this

**Reply by JanDeDobbeleer** [reply]
@PhyAMR it shows the active virtual environment. So the one we display should always be the last activated. They set an environment variable on activation, and we read that one.

**Reply by PhyAMR** [reply]
But if deactivate the virtual enviroment it shouldn't show that I'm on a venv or is anyway to just show when I'm on a venv and when I deactivate it don't like in the regular terminal

**Reply by JanDeDobbeleer** [reply]
@PhyAMR I don't understand that sentence. The python tool you use for managing environments is responsible to also reset the environment variable. If they don't, that's a bug on their end. Also check if the default prompt hijacking isn't enabled by those tools.

**Comment by joadataarg** [comment]
Hi thanks for your input, could you help me to be able to see my surroundings every time I activate it in my power shell console please. When I run conda activate testEnvironmentCreation1 in my terminal does not appear:

![image](https://github.com/user-attachments/assets/60c5dd6f-3437-48a1-bc05-7cb549def1f7)

I have noticed that the environment is not activated

My profile file is as follows:

![image](https://github.com/user-attachments/assets/3dfdde9b-894a-43da-b8d8-73a2ad9d9f66)

**Could you show me how to find the json file, which modification would be the correct one to be able to visualize when activating and deactivating virtual environments. Thank you**

**Reply by JanDeDobbeleer** [reply]
@joaquindev23 please read the docs on how to customize [here](https://www.ohmyposh.dev/docs/installation/customize#custom-configuration). The docs for the python segment can be found [here](https://www.ohmyposh.dev/docs/segments/languages/python) once you have your own custom config.

**Reply by Wontfallo** [reply]
classic Dev answer. took as much time to copy links to docs and tell you to read them instead of just tell you the answer. I mean I get it, trying to teach you to fish right? that way you never starve otherwise if he gives you the fish and if he is not around you would starve? yeah I feel that, except I also feel like if they are here posting a question to a pretty simple answer  that the dev could provide to you, is that user doesn't fish nor do they want to fish.   I see it both ways, but also annoyed at the classic snub nose dev response "RTFD"  how about sniff your own farts, jesus your eyes roll so hard in the back of your skull I thought you might fall off your chair.

**Reply by Wontfallo** [reply]
also thanks for the links.   (seriously)

...fart sniffer

**Reply by Wontfallo** [reply]
@joaquindev23  in your terminal if you run `conda init` then look at your Microsoft.Powershell.profile.ps1 (or whatever the name of your current powershell profile is)  You will likely see Conda added hooks into your ps profile and it would have done so after your Oh-My-Posh. If you see that then simply copy and paste that hook that conda added and put that in the line above the Oh-My-Posh command.  See Below. 

![image](https://github.com/user-attachments/assets/6c99825a-8814-478a-8de8-93e6558c9a65)

 
## Also for anyone else that were told to "read the docs"  

1. **_According to the docs,_** first export the Oh-My-Posh theme you are using and like but just want to edit or tweak, like show which python virtual environment you are currently in if any at all.  [See Below]
`oh-my-posh config export --output ("put path and name of this new theme.json")`

![exportCurrentTheme2NewLocationAndName](https://github.com/user-attachments/assets/2a00a7c4-11ce-4911-97cc-9679ca8f53e9)


2.  **_According to the docs,_**   once you exported your newly named theme json file and opened that new theme that you just created by exporting it-- you are now going to open it in your text editor of choice you will notice this config .json file follows the format in what it calls Block and Segments.  (This if for json only but I assume same with the other two types) [SEE BELOW] In the editor with your newly named config file open find "blocks" and within "blocks" you will find "segments"
![BlocksSegments](https://github.com/user-attachments/assets/24d2bb7f-db2f-4494-931d-20c556a1d00f)


3.  **_According to the docs,_**  Python has it's own segment that you need to add. 


![image](https://github.com/user-attachments/assets/6d6ddc3e-c530-4857-bffe-2dfeec1aa356)


 Simply copy and past the following at the end of the last segment:
[SEE BELOW]

 
```
         ,
        {
          "type": "python",
          "style": "powerline",
          "powerline_symbol": "",
          "foreground": "#100e23",
          "background": "#906cff",
          "template": "  {{ if .Venv }}({{ .Venv }}) {{ end }}{{ .Version }} "
        }
```


![afterExportEditThemeAddPythonSegment](https://github.com/user-attachments/assets/47292bab-e086-4275-b2bb-4e39c3821e05)

## Also notice for each segment in the .json format end with a coma except the last segment.

```
}
  something something 
},
{
  the last squirrelly brace does not get a coma, so make sure its added to WHAT WAS the last segment and NOT on the segment you 
  just add since now that is the last. [see ABOVE] and see this example here
} 
```


4.) ** Thats it just copy and paste the python segment into your new config that you created by exporting it from some config you wanted to tweak.  And then change the location in your powershell profile to point to that new oh-my-posh theme config.json file.**  

***CLOSE all terminal and reopen for changes to take  [SEE BELOW]  ***


**Don't save this edited tweak config.json oh-my-posh theme in the same location as the rest of the oh-my-posh themes it will be over written. I saved my in my user folder but any where is fine, just make sure to point to that location in your powershell profile.ps1**



![image](https://github.com/user-attachments/assets/f5a4cb61-2173-430b-8517-8e186ee87d0c)

**Comment by HamimAhmed007** [comment]
Hello, love you work,
I am using clean-detailed.omp.json theme with slight changes, I added the python segment and changed it a little. It is showing python but not my virtual environment name .
this is the python segment:
```
{
          "type": "python",
          "style": "diamond",
          "leading_diamond": "\ue0b2",
          "properties": {
            "prefix": " \uE235 "
          },
          "trailing_diamond": "\ue0b0",
          "foreground": "#100e23",
          "background": "#906cff",
          "template": " \ue73c {{ if .Error }}{{ .Error }}{{ else }}{{ if .Venv }}{{ .Venv }} {{ end }}{{ .Full }}{{ end }}"
        }
```
This is the promt without activating virtual environment:
![image](https://github.com/user-attachments/assets/e4a8342f-bb4f-46e5-b7e0-edae35d5b0dc)

This is the prompt with virtual environment:
![image](https://github.com/user-attachments/assets/d538215e-4df7-4307-9608-c8c41561e131)

Thank you

**Reply by JanDeDobbeleer** [reply]
@HamimAhmed007 you need to also specify `"fetch_virtual_env": true` as indicated in the [docs](https://ohmyposh.dev/docs/segments/languages/python).

**Reply by HamimAhmed007** [reply]
Isn't that true on default?
Well,I did that. It still does not work.

**Reply by HamimAhmed007** [reply]
It's working with the old code now. The problem was that my virtual name was venv so ohmyposh theme was filtering it out and showing python.

**Reply by JanDeDobbeleer** [reply]
@HamimAhmed007 oh-my-posh debug will show that for you.

**Reply by HamimAhmed007** [reply]
No. It does not. I ran the command "oh-my-posh print debug". I just returned "\[\][DBG]: \[\]".
Well I changed my virtual environment name and it's showing on the promt now.

**Reply by JanDeDobbeleer** [reply]
@HamimAhmed007 that's because it says `oh-my-posh debug` in my post and not `oh-my-posh print debug`.

**Reply by HamimAhmed007** [reply]
Oh sorry.It's fixed now.
And thank you.

**Comment by H6** [comment]
[For Pixi it's configurable under shell in the config.toml](https://pixi.sh/dev/reference/pixi_configuration/#shell)

```
[shell]
change-ps1 = false
```

**Comment by Paciupa** [comment]
Since [v24.18.1](https://github.com/JanDeDobbeleer/oh-my-posh/releases/tag/v24.18.1) oh-my-posh handles it by default. So now I remove my VIRTUAL_ENV_DISABLE_PROMPT env (important to remove it, setting to 0 is not enough).
@JanDeDobbeleer, thanks! Would be nice to update [FAQ](https://ohmyposh.dev/docs/faq/#python-venv-prompt-changes-on-activating-virtual-environment).
